File: cpp/problems/leetcode/0600_non_negative_integers_without_consecutive_ones/solution.cpp
// LeetCode 600: Non-negative Integers Without Consecutive Ones
#include <alg_std.hpp>
#include <bitset>
#include <cassert>
using namespace std;

class Solution {
    string s;
    vector<vector<vector<int>>> memo;

    int dfs(int i, bool tight, bool prev1) {
        if (i == (int)s.size()) return 1;
        int& cell = memo[i][tight][prev1];
        if (cell != -1) return cell;
        int limit = tight ? (s[i] - '0') : 1;
        int total = 0;
        for (int d = 0; d <= limit; ++d) {
            if (d == 1 && prev1) continue;
            total += dfs(i + 1, tight && d == limit, d == 1);
        }
        return cell = total;
    }

public:
    int findIntegers(int n) {
        if (n == 0) return 1;
        s = bitset<32>(n).to_string();
        s = s.substr(s.find('1'));
        memo.assign(s.size(), vector<vector<int>>(2, vector<int>(2, -1)));
        return dfs(0, true, false);
    }
};

int main() {
    assert(Solution().findIntegers(5) == 5);
    assert(Solution().findIntegers(1) == 2);
    assert(Solution().findIntegers(0) == 1);
    assert(Solution().findIntegers(10) == 8);
    cout << "0600 OK" << endl;
    return 0;
}

File: python/problems/leetcode/0600_non_negative_integers_without_consecutive_ones/solution.py
"""LeetCode 600: Non-negative Integers Without Consecutive Ones."""

from __future__ import annotations

from functools import lru_cache


class Solution:
    def findIntegers(self, n: int) -> int:
        s = bin(n)[2:]

        @lru_cache(maxsize=None)
        def dfs(i: int, tight: bool, prev1: bool) -> int:
            if i == len(s):
                return 1
            limit = int(s[i]) if tight else 1
            total = 0
            for d in range(0, limit + 1):
                if d == 1 and prev1:
                    continue
                total += dfs(i + 1, tight and d == limit, d == 1)
            return total

        return dfs(0, True, False)


if __name__ == "__main__":
    assert Solution().findIntegers(5) == 5
    assert Solution().findIntegers(1) == 2
    assert Solution().findIntegers(0) == 1
    assert Solution().findIntegers(10) == 8
    print("0600 OK")
